# Hyperparameter Tuning Practice

## *Data Science Unit 4 Sprint 2 Assignment 3*

# Gridsearch Hyperparameters

In the guided project, you learned how to use sklearn's `GridsearchCV` and `keras-tuner` libraries to tune the hyperparameters of a neural network model. For your module project you'll continue using these two libraries, however we are going to make things a little more interesting for you.

Continue to use TensorFlow Keras & a sample of the [Quickdraw dataset](https://github.com/googlecreativelab/quickdraw-dataset) to build a sketch classification model. The dataset has been sampled to only 10 classes and 10000 observations per class.



**Don't forget to switch to GPU on Colab!**

## 0.1 Imports and installs

In [ ]:
! pip install scikeras

In [ ]:
! pip install keras-tuner

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.1/129.1 kB 1.5 MB/s eta 0:00:00


In [ ]:
# native python libraries imports
import math
from time import time
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# sklearn imports
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV

# keras imports
from keras import Sequential
from keras.layers import Dense
from keras.activations import relu, sigmoid
from keras.optimizers import Adam, SGD
from keras.utils import get_file

# required for compatibility between sklearn and keras
from scikeras.wrappers import KerasClassifier, KerasRegressor

# install keras-tuner
from keras_tuner import RandomSearch, BayesianOptimization
from keras_tuner import HyperParameters

## 0.2 Load quickdraw data

In [ ]:
def load_quickdraw10():
    """
    Fill out this doc string, and comment the code, for practice in writing the kind of code that will get you hired.
    """

    URL_ = "https://github.com/LambdaSchool/DS-Unit-4-Sprint-2-Neural-Networks/blob/main/quickdraw10.npz?raw=true"

    path_to_zip = get_file('./quickdraw10.npz', origin=URL_, extract=False)

    data = np.load(path_to_zip)

    # normalize your image data
    max_pixel_value = 255
    X = data['arr_0']/max_pixel_value
    Y = data['arr_1']

    return train_test_split(X, Y, shuffle=True)

In [ ]:
X_train, X_test, y_train, y_test = load_quickdraw10()

25421363/25421363 [==============================] - 0s 0us/step


In [ ]:
X_train.shape

(75000, 784)

In [ ]:
y_train.shape

(75000,)

_____

# Experiment 1

## Tune Hyperperameters using Enhanced GridsearchCV

We are going to use GridsearchCV again to tune a deep learning model however we are going to add some additional functionality to our gridsearch.

Specifically, we are going to automate the generation of how many nodes to use in a layer and how many layers to use in a model!

By the way, yes, there is a function within a function. Try to not let that bother you. An alternative to this would be to create a class. If you're up for the challenge give it a shot. However, consider this a stretch goal that you come back to after you finish going through this assignment.


### Objective

The objective of this experiment is to show you how to automate the generation of layers and layer nodes for the purposes of gridsearch. <br>
Up until now, we've been manually selecting the number of layers and layer nodes.

In [ ]:
# Function to create model, required for KerasClassifier
def create_model(n_layers,  first_layer_nodes, last_layer_nodes, act_funct ="relu", negative_node_incrementation=True):
    """"
    Returns a compiled keras model

    Parameters
    ----------
    n_layers: int
        number of hidden layers in model
        To be clear, this excludes the input and output layer.

    first_layer_nodes: int
        Number of nodes in the first hidden layer

    last_layer_nodes: int
        Number of nodes in the last hidden layer (this is the layer just prior to the output layer)

     act_funct: string
         Name of activation function to use in hidden layers (this excludes the output layer)

    Returns
    -------
    model: keras object
    """

    def gen_layer_nodes(n_layers, first_layer_nodes, last_layer_nodes, negative_node_incrementation=True):
        """
        Generates and returns the number of nodes in each hidden layer.
        To be clear, this excludes the input and output layer.

        Note
        ----
        Number of nodes in each layer is linearly incremented.
        For example, gen_layer_nodes(5, 500, 100) will generate [500, 400, 300, 200, 100]

        Parameters
        ----------
        n_layers: int
            Number of hidden layers
            This values should be 2 or greater

        first_layer_nodes: int

        last_layer_nodes: int

        Returns
        -------
        layers: list of ints
            Contains number of nodes for each layer
        """

        # throws an error if n_layers is less than 2
        assert n_layers >= 2, "n_layers must be 2 or greater"

        layers = []

        # PROTIP: IF YOU WANT THE NODE INCREMENTATION TO BE SPACED DIFFERENTLY
        # THEN YOU'LL NEED TO CHANGE THE WAY THAT IT'S CALCULATED - HAVE FUN!
        # when set to True number of nodes are decreased for subsequent layers
        # NOTE: the order of the number of nodes doesn't matter
        if negative_node_incrementation:
            # subtract this amount from previous layer's nodes in order to increment towards smaller numbers
            nodes_increment = (last_layer_nodes - first_layer_nodes)/ (n_layers-1)

        # when set to False number of nodes are increased for subsequent layers
        else:
            # add this amount from previous layer's nodes in order to increment towards larger numbers
            nodes_increment = (last_layer_nodes + first_layer_nodes)/ (n_layers-1)

        nodes = first_layer_nodes

        for i in range(1, n_layers+1):

            layers.append(math.ceil(nodes))

            # increment nodes for next layer
            nodes = nodes + nodes_increment

        return layers

    # create model
    model = Sequential()

    n_nodes = gen_layer_nodes(n_layers, first_layer_nodes, last_layer_nodes, negative_node_incrementation)

    for i in range(1, n_layers+1):
        if i==1:
            model.add(Dense(first_layer_nodes, input_dim=X_train.shape[1], activation=act_funct))
        else:
            model.add(Dense(n_nodes[i-1], activation=act_funct))


    # output layer
    model.add(Dense(10, # 10 unit/neurons in output layer because we have 10 possible labels to predict
                    activation='softmax')) # use softmax for a label set greater than 2

    # Compile model
    model.compile(loss='sparse_categorical_crossentropy',
                  optimizer='adam', # adam is a good default optimizer
                  metrics=['accuracy'])

    # do not include model.fit() inside the create_model function
    # KerasClassifier is expecting a complied model
    return model


## 1.1 Explore `create_model`



The helper function `gen_layer_nodes()` which is contained inside `create_model()` <br>
returns a list containing the number of nodes for each successive layer.<br>

Let's check that `gen_layer_nodes()` behaves as expected. <br>
In other words, we'll perform a **Unit Test!**

In [ ]:
def gen_layer_nodes(n_layers, first_layer_nodes, last_layer_nodes, negative_node_incrementation=True):
    """
    Generates and returns the number of nodes in each hidden layer.
    To be clear, this excludes the input and output layer.

    Note
    ----
    Number of nodes in each layer is linearly incremented.
    For example, gen_layer_nodes(5, 500, 100) will generate [500, 400, 300, 200, 100]

    Parameters
    ----------
    n_layers: int
        Number of hidden layers
        This values should be 2 or greater

    first_layer_nodes: int

    last_layer_nodes: int

    Returns
    -------
    layers: list of ints
        Contains number of nodes for each layer
    """

    # throws an error if n_layers is less than 2
    assert n_layers >= 2, "n_layers must be 2 or greater"

    layers = []

    # PROTIP: IF YOU WANT THE NODE INCREMENTATION TO BE SPACED DIFFERENTLY
    # THEN YOU'LL NEED TO CHANGE THE WAY THAT IT'S CALCULATED - HAVE FUN!
    # when set to True number of nodes are decreased for subsequent layers
    # NOTE: the order of the number of nodes doesn't matter
    if negative_node_incrementation:
        # subtract this amount from previous layer's nodes in order to increment towards smaller numbers
        nodes_increment = (last_layer_nodes - first_layer_nodes)/ (n_layers-1)
        #print(f'nodes increment = {nodes_increment}')

    # when set to False number of nodes are increased for subsequent layers
    else:
        # add this amount from previous layer's nodes in order to increment towards larger numbers
        nodes_increment = (last_layer_nodes - first_layer_nodes)/ (n_layers-1)
        #print(f'nodes increment = {nodes_increment}')

    nodes = first_layer_nodes

    for i in range(1, n_layers+1):

        layers.append(math.ceil(nodes))

        # increment nodes for next layer
        nodes = nodes + nodes_increment

    return layers


### `negative_node_incrementation = True`
For this case we want the number of nodes to _decrease_ by a constant number for successive layers. <br>So `first_layer_nodes` must be _larger_ than `last_layer_nodes`

In [ ]:
n_layers = 5
first_layer_nodes = 500
last_layer_nodes = 100
negative_node_incrementation = True
n_nodes = gen_layer_nodes(n_layers, first_layer_nodes, last_layer_nodes, negative_node_incrementation)
print(f'Number of nodes in successive layers: {n_nodes}')

Number of nodes in successive layers: [500, 400, 300, 200, 100]


### `negative_node_incrementation = False`
For this case we want the number of nodes to _increase_ by a constant number for successive layers. <br>So `first_layer_nodes` must be _smaller_ than `last_layer_nodes`

In [ ]:
n_layers = 5
first_layer_nodes = 100
last_layer_nodes = 500
negative_node_incrementation = False
n_nodes = gen_layer_nodes(n_layers, first_layer_nodes, last_layer_nodes, negative_node_incrementation)
print(f'Number of nodes in successive layers: {n_nodes}')

Number of nodes in successive layers: [100, 200, 300, 400, 500]


In [ ]:
# use create_model to create a model

# YOUR CODE HERE
model = create_model(n_layers=10, first_layer_nodes=500, last_layer_nodes=100, act_funct='relu')

In [ ]:
# run model.summary() and make sure that you understand the model architecture that you just built
# Notice in the model summary how the number of nodes have been linearly incremented in decreasing values.
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 500)               392500    
                                                                 
 dense_1 (Dense)             (None, 456)               228456    
                                                                 
 dense_2 (Dense)             (None, 412)               188284    
                                                                 
 dense_3 (Dense)             (None, 367)               151571    
                                                                 
 dense_4 (Dense)             (None, 323)               118864    
                                                                 
 dense_5 (Dense)             (None, 278)               90072     
                                                                 
 dense_6 (Dense)             (None, 234)               6

### Build a model, setting `negative_node_incrementation = False`

Use `create_model` to build a model.

- Set `n_layers = 10`
- Set `first_layer_nodes = 100`
- Set `last_layer_nodes = 500`
- Set `act_funct = "relu"`

In [ ]:
# use create_model to create a model

# YOUR CODE HERE
model = create_model(n_layers=10, first_layer_nodes=100, last_layer_nodes=500, act_funct='relu')

In [ ]:
# run model.summary() and make sure that you understand the model architecture that you just built
# Notice in the model summary how the number of nodes have been linearly incremented in decreasing values.
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_11 (Dense)            (None, 100)               78500     
                                                                 
 dense_12 (Dense)            (None, 145)               14645     
                                                                 
 dense_13 (Dense)            (None, 189)               27594     
                                                                 
 dense_14 (Dense)            (None, 234)               44460     
                                                                 
 dense_15 (Dense)            (None, 278)               65330     
                                                                 
 dense_16 (Dense)            (None, 323)               90117     
                                                                 
 dense_17 (Dense)            (None, 367)              

## 1.2 Create a grid search using `sklearn`

### Hyperparameter search

In [ ]:
# define the grid search parameters
param_grid = {'n_layers': [2, 3],
              'epochs': [3],
              "first_layer_nodes": [500, 300],
              "last_layer_nodes": [100, 50]
             }

In [ ]:
model = KerasClassifier(create_model)

In [ ]:
%%time
# Create Grid Search
grid = GridSearchCV(estimator=model,
                    param_grid=param_grid,
                    n_jobs=-2,
                    verbose=1,
                    cv=3)

grid_result = grid.fit(X_train, y_train)

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")

means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']

for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}")

Fitting 3 folds for each of 8 candidates, totalling 24 fits


ValueError: Invalid parameter first_layer_nodes for estimator KerasClassifier.
This issue can likely be resolved by setting this parameter in the KerasClassifier constructor:
`KerasClassifier(first_layer_nodes=500)`
Check the list of available parameters with `estimator.get_params().keys()`

In [ ]:
best_model = grid_result.best_estimator_

In [ ]:
best_model.get_params()

Ok, now that we've played around a bit with  `create_model`, let's build a  simpler model that we'll use to run gridsearches.

### Build model

Use `create_model` to build a model.

- Set `n_layers = 2`
- Set `first_layer_nodes = 500`
- Set `last_layer_nodes = 100`
- Set `act_funct = "relu"`
- Make sure that `negative_node_incrementation = True`

In [ ]:
# use create_model to create a model

###BEGIN SOLUTION
# use create_model to create a model
model = create_model(n_layers=2, first_layer_nodes=500, last_layer_nodes=100, act_funct='relu', negative_node_incrementation=True)
###END SOLUTION

In [ ]:
# run model.summary() and make sure that you understand the model architecture that you just built
model.summary()

In [ ]:
# define the grid search parameters
param_grid = {'n_layers': [2, 3],
              'epochs': [3],
              "first_layer_nodes": [500, 300],
              "last_layer_nodes": [100, 50]
             }

In [ ]:
model = KerasClassifier(create_model, first_layer_nodes=[500, 300],
                                    last_layer_nodes=[100, 50],
                                    n_layers=[2,3],
                                    epochs=3)

In [ ]:
%%time
# Create Grid Search
grid = GridSearchCV(estimator=model,
                    param_grid=param_grid,
                    n_jobs=-2,
                    verbose=1,
                    cv=3)

grid_result = grid.fit(X_train, y_train)

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")

means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']

for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}")

Fitting 3 folds for each of 8 candidates, totalling 24 fits
Epoch 1/3
1563/1563 [==============================] - 7s 3ms/step - loss: 0.6440 - accuracy: 0.8028
Epoch 2/3
1563/1563 [==============================] - 4s 3ms/step - loss: 0.4378 - accuracy: 0.8684
Epoch 3/3
782/782 [==============================] - 1s 1ms/step
Epoch 1/3
1563/1563 [==============================] - 6s 3ms/step - loss: 0.6488 - accuracy: 0.8018
Epoch 2/3
1563/1563 [==============================] - 5s 3ms/step - loss: 0.4418 - accuracy: 0.8659
Epoch 3/3
782/782 [==============================] - 1s 1ms/step
Epoch 1/3
1563/1563 [==============================] - 6s 3ms/step - loss: 0.6391 - accuracy: 0.8049
Epoch 2/3
1563/1563 [==============================] - 5s 3ms/step - loss: 0.4355 - accuracy: 0.8698
Epoch 3/3
782/782 [==============================] - 1s 1ms/step
Epoch 1/3
1563/1563 [==============================] - 6s 3ms/step - loss: 0.6388 - accuracy: 0.8044
Epoch 2/3
1563/1563 [=================

In [ ]:
best_model = grid_result.best_estimator_

In [ ]:
best_model.get_params()

{'model': <function __main__.create_model(n_layers, first_layer_nodes, last_layer_nodes, act_funct='relu', negative_node_incrementation=True)>,
 'build_fn': None,
 'warm_start': False,
 'random_state': None,
 'optimizer': 'rmsprop',
 'loss': None,
 'metrics': None,
 'batch_size': None,
 'validation_batch_size': None,
 'verbose': 1,
 'callbacks': None,
 'validation_split': 0.0,
 'shuffle': True,
 'run_eagerly': False,
 'epochs': 3,
 'first_layer_nodes': 500,
 'last_layer_nodes': 100,
 'n_layers': 2,
 'class_weight': None}

-----

# Experiment 2: Run the Gridsearch Algorithms

In this section, we are going to use the same model and dataset in order to benchmark 3 different gridsearch approaches:

- Gridsearch
- Random Search
- Bayesian Optimization.


Our goal in this experiment is two-fold. We want to see which appraoch

- Scores the highest accuracy
- Has the shortest run time

We want to see how these 3 gridsearch approaches handle these trade-offs and to give you a sense of those trades offs.

### Trade-offs

`Gridsearch` will train a model on every single unique hyperparameter combination, this guarantees that you'll get the highest possible accuracy from your parameter set but your gridsearch might have a very long run-time.

`Random Search` will randomly sample from your parameter set which, depending on how many samples, the run-time might be significantly cut down but you might or might not sample the parameters that correspond to the heightest possible accuracies.

`Bayesian Optimization` has a bit of intelligence built into it's search algorithm but you do need to manually select some parameters which may greatly influence the model learning outcomes.

-------
### Build our hyperparameter dictionary

In [ ]:
# build out our hyperparameter dictionary
hp = HyperParameters()
hp.Int('units', min_value=32, max_value=512, step=32)
hp.Choice('learning_rate',values=[1e-1, 1e-2, 1e-3])
hp.Choice('activation',values=["relu", "sigmoid"])

'relu'

---------
## 2.1 Gridsearch Optimization


### Populate a `sklearn` compatible parameter dictionary

In [ ]:
# build out our hyperparameter dictionary
hyper_parameters = {
    # BUG Fix: cast array as list otherwise GridSearchCV will throw error
    "units": np.arange(32, 512, 32).tolist(),
    "learning_rate": [1e-1, 1e-2, 1e-3],
    "activation":["relu", "sigmoid"]
}

In [ ]:
units = [32,
  64,
  96,
  128,
  160,
  192,
  224,
  256,
  288,
  320,
  352,
  384,
  416,
  448,
  480]

In [ ]:
hyper_parameters

{'units': [32,
  64,
  96,
  128,
  160,
  192,
  224,
  256,
  288,
  320,
  352,
  384,
  416,
  448,
  480],
 'learning_rate': [0.1, 0.01, 0.001],
 'activation': ['relu', 'sigmoid']}

### Build a `sklearn` compatible model function

In [ ]:
def build_model(units, learning_rate, activation):
    """
    Returns a complie keras model ready for keras-tuner gridsearch algorithms
    """

    model = Sequential()

    # hidden layer
    model.add(Dense(units, activation=activation))

    # output layer
    model.add(Dense(10, activation='softmax'))

    model.compile(
        optimizer=Adam(learning_rate),
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy'])

    return model

In [ ]:
gen_layer_nodes

### Apply the "wrapper" to make the model compatible with `sklearn`

In [ ]:
model = KerasClassifier(build_fn=build_model)

In [ ]:
y_train.shape

(75000,)

In [ ]:
# save start time
start = time()

# Create Grid Search
grid = GridSearchCV(estimator=model,
                    param_grid=hyper_parameters,
                    n_jobs=-2,
                    verbose=1,
                    cv=3,
                    error_score='raise')

grid_result = grid.fit(X_train, y_train)

# save end time
end = time()

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")

means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']

for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}")

Fitting 3 folds for each of 90 candidates, totalling 270 fits


ValueError: Invalid parameter activation for estimator KerasClassifier.
This issue can likely be resolved by setting this parameter in the KerasClassifier constructor:
`KerasClassifier(activation=relu)`
Check the list of available parameters with `estimator.get_params().keys()`

In [ ]:
# total run time
total_run_time_in_miniutes = (end - start)/60
total_run_time_in_miniutes

In [ ]:
grid_result.best_params_

In [ ]:
# because all other optimization approaches are reporting test set score
# let's calculate the test set score in this case
best_model = grid_result.best_estimator_
test_acc = best_model.score(X_test, y_test)

In [ ]:
test_acc

 ### Results

Identify and write the the best performing hyperparameter combination and model score.



Hyperparameters:
units: 480
learning_rate: 0.001
activation: relu
Score: 0.8510800004005432


------
## 2.2 Random Search with `keras-tuner`

Be sure to check out the [**docs for Keras-Tuner**](https://keras-team.github.io/keras-tuner/documentation/tuners/). Here you can read about the input parameters for the `RandomSearch` tuner.

### Build model
---

In [ ]:
# because gridsearching can take a lot of time and we are bench marking 3 different approaches
# let's build a simple model to minimize run time

def build_model(hp):
    """
    Returns a complied keras model ready for keras-tuner gridsearch algorithms
    """

    model = Sequential()

    # hidden layer
    model.add(Dense(units=hp.get('units'),activation=hp.get("activation")))

    # output layer
    model.add(Dense(10, activation='softmax'))

    model.compile(
        optimizer=Adam(hp.get('learning_rate')),
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy'])

    return model

In [ ]:
# how many unique hyperparameter combinations do we have?
# HINT: take the product of the number of possible values for each hyperparameter
# save your answer to n_unique_hparam_combos

# YOUR CODE HERE
n_values_in_units = 512/32
n_values_in_learning_rate = 3
n_values_in_activation = 2

n_unique_hparam_combos = n_values_in_units *n_values_in_learning_rate * n_values_in_activation
n_unique_hparam_combos

96.0

In [ ]:
# how many of these do we want to randomly sample?
# let's pick 25% of n_unique_hparam_combos param combos to sample
# save this number to n_param_combos_to_sample

# YOUR CODE HERE
fraction_to_sample = 0.25
n_param_combos_to_sample = int(n_unique_hparam_combos * fraction_to_sample)
n_param_combos_to_sample

24

### Instantiate a `RandomSearch()` object for your grid search

In [ ]:
random_tuner = RandomSearch(
    build_model,
    objective='val_accuracy',
    max_trials=n_param_combos_to_sample, # number of times to sample the parameter set and build a model
    seed=1234,
    hyperparameters=hp, # pass in our hyperparameter dictionary
    directory='./keras-tuner-trial',
    project_name='random_search'
)

In [ ]:
 # take note of Total elapsed time in print out -- took ~10 minutes without GPU
random_tuner.search(
    X_train,
    y_train,
    epochs=3,
    validation_data=(X_test, y_test),
)

Trial 24 Complete [00h 00m 42s]
val_accuracy: 0.6878399848937988

Best val_accuracy So Far: 0.8748800158500671
Total elapsed time: 00h 14m 27s


In [ ]:
# identify the best score and hyperparamter (should be at the top since scores are ranked)
random_tuner.results_summary()

Results summary
Results in ./keras-tuner-trial/random_search
Showing 10 best trials
Objective(name="val_accuracy", direction="max")

Trial 06 summary
Hyperparameters:
units: 416
learning_rate: 0.001
activation: relu
Score: 0.8748800158500671

Trial 18 summary
Hyperparameters:
units: 256
learning_rate: 0.001
activation: relu
Score: 0.8723999857902527

Trial 13 summary
Hyperparameters:
units: 384
learning_rate: 0.001
activation: relu
Score: 0.8708400130271912

Trial 22 summary
Hyperparameters:
units: 288
learning_rate: 0.001
activation: relu
Score: 0.8652799725532532

Trial 00 summary
Hyperparameters:
units: 512
learning_rate: 0.001
activation: sigmoid
Score: 0.8639600276947021

Trial 19 summary
Hyperparameters:
units: 192
learning_rate: 0.001
activation: sigmoid
Score: 0.8551599979400635

Trial 04 summary
Hyperparameters:
units: 96
learning_rate: 0.001
activation: sigmoid
Score: 0.8404399752616882

Trial 07 summary
Hyperparameters:
units: 224
learning_rate: 0.01
activation: sigmoid
Scor

 ### Results

Identify and write the the best performing hyperparameter combination and model score.
Note that because this is Random Search, multiple runs might have slighly different outcomes.



Trial 06 summary
Hyperparameters:
units: 416
learning_rate: 0.001
activation: relu
Score: 0.8748800158500671

------
## 2.3 Bayesian Optimization with `keras-tuner`

![](https://upload.wikimedia.org/wikipedia/commons/0/02/GpParBayesAnimationSmall.gif)

Be sure to check out the [**docs for Keras-Tuner**](https://keras-team.github.io/keras-tuner/documentation/tuners/). Here you can read about the input parameters for the `BayesianOptimization` tuner.

Pay special attention to these `BayesianOptimization` parameters: `num_initial_points` and `beta`.

`num_initial_points`:

Number of randomly selected hyperparameter combinations to try before applying bayesian probability to determine liklihood of which param combo to try next based on expected improvement


`beta`:

Larger values means more willing to explore new hyperparameter combinations (analogous to searching for the global minimum in Gradient Descent), smaller values means that it is less willing to try new hyperparameter combinations (analogous to getting stuck in a local minimum in Gradient Descent).

As a start, error on the side of larger values. What defines a small or large value you ask? That question would pull us into the mathematical intricacies of Bayesian Optimization and Gaussian Processes. For simplicity, notice that the default value is 2.6 and work from there.

In [ ]:
# we know that 24 samples is about 25% of 96 possible hyper-parameter combos
# let's set up a run with the same parameters we used for RandomSearch() so the comparison will be aplles-to-apples
# feel free to play with any of these numbers later
max_trials=24
num_initial_points=5
beta=5.0

#### Instantiate a `BayesianOptimization()` object for your grid search

In [ ]:
bayesian_tuner = BayesianOptimization(
    build_model,
    objective='val_accuracy',
    max_trials=max_trials,
    hyperparameters=hp, # pass in our hyperparameter dictionary
    num_initial_points=num_initial_points,
    beta=beta,
    seed=1234,
    directory='./keras-tuner-trial',
    project_name='bayesian_optimization_4',
)

In [ ]:
bayesian_tuner.search(
    X_train,
    y_train,
    epochs=3,
    validation_data=(X_test, y_test),
)

Trial 24 Complete [00h 00m 42s]
val_accuracy: 0.8748000264167786

Best val_accuracy So Far: 0.8763599991798401
Total elapsed time: 00h 14m 06s


In [ ]:
bayesian_tuner.results_summary()

Results summary
Results in ./keras-tuner-trial/bayesian_optimization_4
Showing 10 best trials
Objective(name="val_accuracy", direction="max")

Trial 09 summary
Hyperparameters:
units: 448
learning_rate: 0.001
activation: relu
Score: 0.8763599991798401

Trial 21 summary
Hyperparameters:
units: 512
learning_rate: 0.001
activation: relu
Score: 0.875

Trial 23 summary
Hyperparameters:
units: 512
learning_rate: 0.001
activation: relu
Score: 0.8748000264167786

Trial 06 summary
Hyperparameters:
units: 512
learning_rate: 0.001
activation: relu
Score: 0.874239981174469

Trial 20 summary
Hyperparameters:
units: 512
learning_rate: 0.001
activation: relu
Score: 0.8740800023078918

Trial 22 summary
Hyperparameters:
units: 512
learning_rate: 0.001
activation: relu
Score: 0.8724799752235413

Trial 16 summary
Hyperparameters:
units: 320
learning_rate: 0.001
activation: relu
Score: 0.8715999722480774

Trial 19 summary
Hyperparameters:
units: 512
learning_rate: 0.001
activation: relu
Score: 0.870079994

 ### Results

Identify and write the the best performing hyperparameter combination and model score.
Note that because this is  Bayesian Optimization, multiple runs might have slighly different outcomes.



Trial 09 summary
Hyperparameters:
units: 448
learning_rate: 0.001
activation: relu
Score: 0.8763599991798401

We should point out that Gridsearch split the training set internally and created a test set whereas keras-tuner allows us to pass in a test set. This means that the keras-tuner algorithms were using one test set and our skearn GridSearchCV was using a different test set - so this isn't a perfectly exact 1-to-1 comparision but it'll have to do. In order to compensate for this, we did score the best model on the same test set that keras-tuner used.

_______

# Conclusion

The spirit of this experiment is to expose you to the idea of benchmarking and comparing the trade-offs of various gridsearch approaches.

Even if we did find a way to pass in the original test set into GridSearchCV, we can see that both Random Search and Bayesian Optimization are arguably better alternatives to a brute force grid search when we consider the trade-offs of run time and locating the best performing model.

----

# Stretch Goals

- Feel free to run whatever gridsearch experiments on whatever models you like!

In [ ]:
# this is your open playground - be free to explore as you wish
